docs/design/docker-deploy.md

```diff
@@ -0,0 +1,651 @@
+# Docker Deploy - Design
+
+## Abstract
+
+Add convenience for users to containerize and deploy their Angular application via Docker.
+
+Provide tasks for common Docker workflows:
+
+* Generate starter `Dockerfile` and `docker-compose.yml` files.
+* Build and Push an Angular app image to a Docker Registry.
+* Deploy and run an Angular app container in different Docker Machine environments.
+
+
+## Requirements
+
+1. Requires user to have Docker CLI tools installed.
+   (See also: ["Implementation Approaches"](#implementation-approaches))
+1. User is free to use the Angular CLI without Docker. By default, do not generate Docker files upon creation of a new project (`ng new`, `ng init`).
+1. Don't recreate the wheel. Implement the common Docker use cases that make it convenient for Angular applications.
+1. Don't inhibit users from using the standalone Docker CLI tools for other use cases.
+1. Assumes 1:1 Dockerfile with Angular project. Other services should be outside (or sibling to) the ng app project folder.
+1. Generated starter Dockerfile will use the current official Node.js LTS base image.
+1. Generated Dockerfile will use `npm run docker-start` as its default CMD. Requires new `package.json` script.
+1. User is free to modify and customize all generated files directly without involvement by the Angular CLI.
+1. Image builds will support all Docker build options.
+1. Container deploys will support all Docker run options.
+1. Deploying to a Docker Machine can be local or remote.
+1. Deploys can be made to different environments (dev, stage, prod) on the same or different Docker Machines.
+1. Image pushes can be made to Docker Hub, AWS ECR, and other public/private registries.
+1. Adhere to [Docker security best practices](https://docs.docker.com/engine/security/).
+1. Use sensible defaults to make it easy for users to get started.
+1. Utilize ngConfig project config file to save information for successive `docker:*` command runs.
+1. Support `--dry-run` and `--verbose` flags.
+
+
+## Proposed CLI API
+
+### Overview
+
+Initialize Docker files and configuration for a default build and run environment:
```

Comment:
Alternatively, we could have `ng docker init`, `ng docker log`, `ng docker deploy`, ...